helix-term/src/ui/editor.rs

```diff
@@ -695,6 +757,342 @@ impl EditorView {
+                return Some(
+                    nodes
+                        .iter()
+                        .take(visual_cursor_pos as usize)
+                        .map(|elem| elem.clone())
+                        .collect(),
+                );
+            }
+        }
+
+        let context_nodes = doc
+            .language_config()
+            .and_then(|lang| lang.context_query())?;
+
+        let start_index = context_nodes.query.capture_index_for_name("context")?;
+        let end_index = context_nodes
+            .query
+            .capture_index_for_name("context.end")
+            .unwrap_or(start_index);
+
+        // context is list of numbers of lines that should be rendered in the LSP context
+        let mut context: HashSet<StickyNode> = HashSet::new();
+
+        let mut cursor = QueryCursor::new();
+
+        // only run the query from start to the cursor location
+        cursor.set_byte_range(0..last_scan_byte);
+        let query = &context_nodes.query;
+        let query_nodes = cursor.matches(query, tree.root_node(), RopeProvider(text));
+
+        for matched_node in query_nodes {
+            // find @context.end nodes
+            let node_byte_range = Self::get_context_paired_range(
+                &matched_node,
+                start_index,
+                end_index,
+                top_first_byte,
+                last_scan_byte,
+            );
+
+            for node in matched_node.nodes_for_capture_index(start_index) {
+                if (!node.byte_range().contains(&last_scan_byte)
+                    || !node.byte_range().contains(&top_first_byte))
+                    && node.start_position().row != anchor_line + context.len()
+                    && node_byte_range.is_none()
+                {
+                    continue;
+                }
+
+                context.insert(StickyNode {
+                    line: node.start_position().row,
+                    visual_line: 0,
+                    byte_range: node_byte_range
+                        .as_ref()
+                        .unwrap_or(&(node.start_byte()..node.start_byte()))
+                        .clone(),
+                    indicator: None,
+                    anchor: view.offset.anchor,
+                    has_context_end: node_byte_range.is_some(),
+                });
+            }
+        }
+        // context should be filled by now
+        if context.is_empty() {
+            return None;
+        }
+
+        // always cap the maximum amount of sticky contextes to 1/3 of the viewport
+        // unless configured otherwise
+        let max_lines = config.sticky_context.max_lines;
+        let max_nodes_amount = if max_lines == 0 {
+            viewport.height as usize / 3
+        } else {
+            max_lines.min(viewport.height) as usize
+        };
+
+        let mut result: Vec<StickyNode> = context.into_iter().collect();
+        result.sort_by(|lnode, rnode| lnode.line.cmp(&rnode.line));
```

Comment:
You could also directly implement `PartialOrd` for `StickNode` of course

Reply:
The main reason why i went with a `HashSet` here is to ensure uniqueness.

Considering that my nodes usually shouldn't exceed a limit in which the sorting would actually matter, i think the `BTree` with the additional impls would be overkill here.

Reply:
BTreeSet also ensures uniqueness.

The problem is that `StickyNode` is a relatively big object and so allocating a vec only to immediately discard it is wasteful.

This will run a lot, pretty much each time text changes, so avoiding small allocations could help, especially outside of Linux :)

Reply:
To be fail you don't need uniquness while building the set. In that case pushing everything to a vec and `sort_unstable` plus `deadup` is faster. A `BTreeSet` is only worth it if you need to maintain sorted order and uniquess while modifying the set. CPUs are just really good at working with contiguous data

Reply:
> To be fail you don't need uniquness while building the set. In that case pushing everything to a vec and `sort_unstable` plus `deadup` is faster. A `BTreeSet` is only worth it if you need to maintain sorted order and uniquess while modifying the set. CPUs are just really good at working with contiguous data

Interestingly, you're right with that i don't need uniqueness anymore.
I think the `HashSet` was a reminent of the old implementation, where i was just iteration from the Top position all the way up to the topmost tree-sitter node.
Fix will be pushed.

Reply:
Actually, i do need to ensure uniqueness @pascalkuthe as i just noticed the following bug:
<img width="1800" alt="Screenshot 0005-07-28 at 19 40 23" src="https://github.com/helix-editor/helix/assets/13474089/4711d876-64ac-4a56-bf38-0cc0ef4effea">
Not sure if that's an issue of my query, but even if, i think we *should* ensure that this isn't happening.

Reply:
I mean deduplication is fine but you can use `Vec::deadup` for that and it will be faster than a `btreemap` or `hashset` (if you need a sorted list anyway as in this case)

Reply:
`Vec::dedup` only works correctly if the `Vec` is already sorted, which is not a guarantee we have here I think ?

Reply:
> `Vec::dedup` only works correctly if the `Vec` is already sorted, which is not a guarantee we have here I think ?

we have, as i can sort the Vector before deduping it.

Reply:
Yeag and sorting a Vec once is faster than inserting into a `BTreeSet` repeatedly. Btw please use `unstable_sort` here it doesn't make a difference (since you deduplicate anyways) and is quite a bit faster